Sample app improvements (runtime permissions, disposable handling, refactoring)

lukaszkalnik opened the pull request:
Add a runtime ACCESS_COARSE_LOCATION permission check and request for Android 6 devices to the sample app.
Also extract ble scan error handling to a separate helper class and extract error message strings to `strings.xml`.
Add disposable handling by disposing subscriptions in `onPause()`. Add some missing annotations.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Polidea/RxAndroidBle?pullRequest=501) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Polidea/RxAndroidBle?pullRequest=501) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Polidea/RxAndroidBle?pullRequest=501) it.</sub>

lukaszkalnik commented:
I have entered my first and last name while signing the CLA and it seems to have been misinterpreted as a second contributor (`Lukasz Kalnik`). Could you please remove it? Thanks.

dariuszseweryn commented on sample/src/main/java/com/polidea/rxandroidble2/sample/util/LocationPermission.java:
```diff
@@ -0,0 +1,44 @@
+        // Utility class
+    }
+
+    private static int REQUEST_PERMISSION_COARSE_LOCATION = 9358;
```
```suggestion
    private static final int REQUEST_PERMISSION_COARSE_LOCATION = 9358;
```

dariuszseweryn commented on sample/src/main/java/com/polidea/rxandroidble2/sample/util/ScanExceptionHandler.java:
```diff
@@ -0,0 +1,104 @@
+            if (resId != null) {
+                text = context.getString(resId);
+            } else {
+                // unknown error - return default message
```
An additional warning could be useful here as well I suppose

dariuszseweryn replied:
```suggestion
                // unknown error - return default message
                Log.w("Scanning", String.format("No message found for reason=%d. Consider adding one.", reason));
```
I was actually thinking about giving a warning log that we got a `reason` that has no error message so it will be clear what to fix

dariuszseweryn commented:
I cannot remove`Lukasz Kalnik` from contributors as one of your commits (`1a92c6e17d14751d60cfe1fc1ccf8972ff164f69`) was made with that user who cannot be attributed to your github account. To fix this situation you would need to change this commit's author to the proper one (`lukaszkalnik`)

lukaszkalnik commented:
Actually this commit doesn't change anything in perspective of the whole PR (just changes two Strings in two Activities, which I later moved anyway to strings.xml in another commit).
Can you remove it from history?

dariuszseweryn commented:
@lukaszkalnik Sorry for the delay — I was on a conference and short holiday. Actually you may and should use a force push to get rid of the commit by `Lukasz Kalnik` from this PR (afterall this PR is from your fork and you do have full control over it which I probably have not). Alternatively you may close this PR and create another one with different commits in it

lukaszkalnik commented:
I rebased and removed the commit (and force pushed) but the offending commit is still there... There is now just one merge remote commit on top.
Also when I created new PR from the same branch, the commit was still in the history, although I don't see it locally anymore.
Probably the best would be to just create a new branch from cherry-picked commits.

dariuszseweryn commented:
I have checkout your repo and I see this commit whose author is `lukasz.kalnik@grandcentrix.net` where the rest of your commits are from `lukasz.kalnik@gmail.com` so the history of your commits does not seem to be overwritten with via `push -f`. Quickly searched google and there seem to be [a fairly straightforward instruction on how you can change the author](https://stackoverflow.com/questions/3042437/how-to-change-the-commit-author-for-one-specific-commit) — I hope this helps

lukaszkalnik commented:
Thank you for your suggestions. Seems like the rebase in Source Tree doesn't work so well. I did it from the command line now and it worked :)

dariuszseweryn commented on sample/src/main/java/com/polidea/rxandroidble2/sample/example2_connection/ConnectionExampleActivity.java:
```diff
@@ -122,4 +128,12 @@ private void updateUI() {
+        super.onPause();
+
+        triggerDisconnect();
+        compositeDisposable.dispose();
```
```suggestion
        compositeDisposable.clear();
```
Otherwise the next time the activity will resume nothing will work

dariuszseweryn commented on sample/src/main/java/com/polidea/rxandroidble2/sample/example4_characteristic/CharacteristicOperationExampleActivity.java:
```diff
@@ -203,4 +214,10 @@ private boolean hasProperty(BluetoothGattCharacteristic characteristic, int prop
+    @Override
+    protected void onPause() {
+        super.onPause();
+        compositeDisposable.dispose();
```
```suggestion
        compositeDisposable.clear();
```

dariuszseweryn commented on sample/src/main/java/com/polidea/rxandroidble2/sample/util/ScanExceptionHandler.java:
```diff
@@ -0,0 +1,104 @@
+            }
+        }
+
+        Log.w("EXCEPTION", text, exception);
```
```suggestion
        Log.w("Scanning", text, exception);
```
Everyone will know that this is an exception because of a stacktrace ;)
